src/sentry/integrations/github/blame.py

```diff
@@ -0,0 +1,213 @@
+import logging
+from dataclasses import asdict
+from datetime import timezone
+from typing import Any, Dict, List, Optional, Sequence, TypedDict, Union
+
+from django.utils.datastructures import OrderedSet
+from isodate import parse_datetime
+
+from sentry.integrations.mixins.commit_context import CommitInfo, FileBlameInfo, SourceLineInfo
+from sentry.shared_integrations.response.mapping import MappingApiResponse
+
+logger = logging.getLogger("sentry.integrations.github")
+
+
+class GitHubAuthor(TypedDict):
+    name: Optional[str]
+    email: Optional[str]
+
+
+class GitHubFileBlameCommit(TypedDict):
+    oid: str
+    author: Optional[GitHubAuthor]
+    message: str
+    committedDate: str
+
+
+class GitHubFileBlameRange(TypedDict):
+    commit: GitHubFileBlameCommit
+    startingLine: int
+    endingLine: int
+    age: int
+
+
+FilePathMapping = Dict[str, Dict[str, OrderedSet]]
+
+
+def generate_file_path_mapping(
+    files: Sequence[SourceLineInfo],
+) -> FilePathMapping:
+    """
+    Generates a nested mapping of repo -> ref -> file paths.
+    This map is used to dedupe matching repos, refs, and file paths and only query
+    for them once. This mapping is passed to `create_blame_query` and
+    `extract_commits_from_blame_response`.
+    """
+    file_path_mapping: FilePathMapping = {}
+    for file in files:
+        repo = file_path_mapping.setdefault(file.repo.name, {})
+        paths = repo.setdefault(file.ref, OrderedSet())
+        paths.add(file.path)
+
+    return file_path_mapping
+
+
+def create_blame_query(file_path_mapping: FilePathMapping) -> str:
+    """
+    Create a GraphQL query to get blame information for a list of files
+    """
+    repo_queries = ""
+    for repo_index, (full_repo_name, ref) in enumerate(file_path_mapping.items()):
+        ref_queries = ""
+        for ref_index, (ref_name, file_paths) in enumerate(ref.items()):
+            blame_queries = "".join(
+                [
+                    _make_blame_query(file_path, file_path_index)
+                    for file_path_index, file_path in enumerate(file_paths)
+                ]
+            )
+            ref_queries += _make_ref_query(ref_name, blame_queries, ref_index)
+
+        try:
+            [repo_owner, repo_name] = full_repo_name.split("/", maxsplit=1)
+        except ValueError:
+            continue
+
+        repo_queries += _make_repo_query(repo_name, repo_owner, ref_queries, repo_index)
+
+    return f"""query {{{repo_queries}\n}}"""
+
+
+def extract_commits_from_blame_response(
```

Comment:
Good thinking - switched to `logger.error`. Will probably end up silencing them later since it is expected to have mappings set up wrong, but it's good to monitor initially.